### PR TITLE
support virtual ab ota

### PR DESCRIPTION
**Pull request by ceiba1985**
1. Add some code to support virtual ab ota.
2. Sync some code with bootctrl.

Tracked-On: OAM-100724
Signed-off-by: ji, zhenlong z <zhenlong.z.ji@intel.com>

### CODE REVIEW

File: libfastboot/fastboot.c [hunk 1]
```diff
@@ -799,6 +826,12 @@ static void cmd_flash(INTN argc, CHAR8 **argv)
 		fastboot_fail("Allocation error");
 		return;
 	}
+
+    if (!can_erase_or_flash_partition(label)) {
+        fastboot_fail("Currently virtual a/b ota is in progress...");
```
Comment (GangSecurity) [inline libfastboot/fastboot.c:831]: use TAB rather than 4 spaces in this file

File: libfastboot/fastboot.c [hunk 2]
```diff
@@ -423,6 +424,32 @@ const char* fastboot_slot_get_active()
...
+    ab_ops.write_ab_metadata = avb_ab_data_write;
+
+    ops = uefi_avb_ops_new();
+    if (ops == NULL) {
+        error(L"Error allocating AvbOps when slot_init.");
+        return 0;
+    }
+    ab_ops.ops = ops;
+
+
```
Comment (GangSecurity) [inline libfastboot/fastboot.c:441]: remove blank line

File: libfastboot/fastboot.c [hunk 3]
```diff
@@ -423,6 +424,32 @@ const char* fastboot_slot_get_active()
...
 }
 
+static int can_erase_or_flash_partition(CHAR16 *label) {
+    struct AvbABOps ab_ops;
+    AvbOps *ops;
+
+    ab_ops.read_ab_metadata = avb_ab_data_read;
+    ab_ops.write_ab_metadata = avb_ab_data_write;
+
+    ops = uefi_avb_ops_new();
```
Comment (GangSecurity) [inline libfastboot/fastboot.c:434]: where to free ops?

File: libfastboot/fastboot.c [hunk 4]
```diff
@@ -423,6 +424,33 @@ const char* fastboot_slot_get_active()
 	return p;
 }
 
+static int can_erase_or_flash_partition(CHAR16 *label)
+{
+        struct AvbABOps ab_ops;
```
Comment (GangSecurity) [inline libfastboot/fastboot.c:429]: use TAB instead of 4 or 8 spaces

**Review by GangSecurity — APPROVED** [review]

**Review by kwang13 — APPROVED** [review]

**Pull request merged**